core/request.go

```diff
@@ -121,19 +121,133 @@ func httpTransfer(c CatalogEntry, t *TransferRequest) (string, error) {
 	return r.Pfn, nil
 }
 
+// Store returns a Decorator that stores request
+func Store() Decorator {
+	return func(r Request) Request {
+		return RequestFunc(func(t *TransferRequest) error {
+			t.Id = time.Now().Unix()
+			item := &Item{
+				Value:    *t,
+				priority: t.Priority,
+			}
+			fmt.Println(*t)
+			err := TFC.InsertRequest(*t)
+			if err != nil {
+				return err
+			} else {
+				heap.Push(&RequestQueue, item)
+			}
+			log.WithFields(log.Fields{
+				"Request": t,
+			}).Println("Request Saved")
+			return r.Process(t)
+		})
+	}
+}
+
+// Delete returns a Decorator that deletes request from heap
+func Delete() Decorator {
+	return func(r Request) Request {
+		return RequestFunc(func(t *TransferRequest) error {
+			// Delete request from PriorityQueue. The complexity is O(n) where n = heap.Len()
+			index := -1
+			var err error
+
+			for _, item := range RequestQueue {
+				if item.Value.Id == t.Id {
+					index = item.index
+					break
+				}
+			}
+
+			if index < RequestQueue.Len() && index >= 0 {
+				err = TFC.UpdateRequest(t.Id, "deleted")
+				if err != nil {
+					t.Status = "error"
+					return err
+				} else {
+					// TODO: May be we need to add lock over here.
+					heap.Remove(&RequestQueue, index)
+					t.Status = "deleted"
+					log.WithFields(log.Fields{
+						"Request": t,
+					}).Println("Request Deleted")
+				}
+			} else {
+				t.Status = "error"
+				err = errors.New("Can't find request in heap")
+				return err
+			}
+
+			return r.Process(t)
+		})
+	}
+}
+
+// Transfer returns a Decorator that performs request transfers by pull model
+func PullTransfer() Decorator {
+	return func(r Request) Request {
+		return RequestFunc(func(t *TransferRequest) error {
+			log.WithFields(log.Fields{
+				"Request": t.String(),
+			}).Println("Request Transfer")
+			// obtain information about source and destination agents
+			url := fmt.Sprintf("%s/status", t.DstUrl)
+			resp := utils.FetchResponse(url, []byte{})
+			if resp.Error != nil {
+				return resp.Error
+			}
+			var dstAgent AgentStatus
+			err := json.Unmarshal(resp.Data, &dstAgent)
+			if err != nil {
+				return err
+			}
+			url = fmt.Sprintf("%s/status", t.SrcUrl)
+			resp = utils.FetchResponse(url, []byte{})
+			if resp.Error != nil {
+				return resp.Error
+			}
+			var srcAgent AgentStatus
+			err = json.Unmarshal(resp.Data, &srcAgent)
+			if err != nil {
+				return err
+			}
+			// if both are up then send acknowledge message to destination on /pullack url.
+			body, err := json.Marshal(t)
+			if err != nil {
+				return err
+			}
+			url = fmt.Sprintf("%s/pull", t.DstUrl)
```

Comment:
In PushTransfer code we have possibility to switch to middleware tools. The PullTransfer should do the same. We should structure a code such a way that we can switch to middleware (and we will test this) if necessary.